test/extensions/config_subscription/grpc/xds_failover_integration_test.cc

```diff
@@ -906,4 +906,149 @@ TEST_P(XdsFailoverAdsIntegrationTest,
+  EXPECT_TRUE(compareDiscoveryRequest(CdsTypeUrl, "", {}, {}, {}, true,
+                                      Grpc::Status::WellKnownGrpcStatus::Ok, "",
+                                      failover_xds_stream_.get()));
+  sendDiscoveryResponse<envoy::config::cluster::v3::Cluster>(
+      CdsTypeUrl, {ConfigHelper::buildCluster("failover_cluster_0")},
+      {ConfigHelper::buildCluster("failover_cluster_0")}, {}, "failover1", {},
+      failover_xds_stream_.get());
+  // Wait for an EDS request, and send its response.
+  test_server_->waitForGaugeEq("cluster_manager.warming_clusters", 1);
+  test_server_->waitForGaugeEq("cluster.failover_cluster_0.warming_state", 1);
+  // Ensure basic flow with failover works.
+  EXPECT_TRUE(compareDiscoveryRequest(
+      EdsTypeUrl, "", {"failover_cluster_0"}, {"failover_cluster_0"}, {}, false,
+      Grpc::Status::WellKnownGrpcStatus::Ok, "", failover_xds_stream_.get()));
+  sendDiscoveryResponse<envoy::config::endpoint::v3::ClusterLoadAssignment>(
+      EdsTypeUrl, {buildClusterLoadAssignment("failover_cluster_0")},
+      {buildClusterLoadAssignment("failover_cluster_0")}, {}, "failover1", {},
+      failover_xds_stream_.get());
+  test_server_->waitForGaugeEq("cluster_manager.warming_clusters", 0);
+  test_server_->waitForGaugeGe("cluster_manager.active_clusters", 2);
+  test_server_->waitForGaugeEq("cluster.failover_cluster_0.warming_state", 0);
+  EXPECT_EQ(2, test_server_->gauge("control_plane.connected_state")->value());
+  EXPECT_TRUE(compareDiscoveryRequest(CdsTypeUrl, "failover1", {}, {}, {}, false,
+                                      Grpc::Status::WellKnownGrpcStatus::Ok, "",
+                                      failover_xds_stream_.get()));
+  EXPECT_TRUE(compareDiscoveryRequest(LdsTypeUrl, "", {}, {}, {}, false,
+                                      Grpc::Status::WellKnownGrpcStatus::Ok, "",
+                                      failover_xds_stream_.get()));
+
+  // Envoy has received CDS and EDS responses, it means the failover is available.
+  // Now disconnect the failover source, this should result in an LDS failure.
+  // After that add a notification to the main thread to ensure that the retry timer kicks in.
+  failover_xds_stream_->finishGrpcStream(Grpc::Status::Internal);
+  test_server_->waitForCounterGe("listener_manager.lds.update_failure", 1);
+  absl::Notification notification;
+  test_server_->server().dispatcher().post([&]() { notification.Notify(); });
+  notification.WaitForNotification();
+  timeSystem().advanceTimeWait(std::chrono::milliseconds(1000));
+
+  // In this case (received a response), both EnvoyGrpc and GoogleGrpc keep the connection open.
+  result = failover_xds_connection_->waitForNewStream(*dispatcher_, failover_xds_stream_);
+  RELEASE_ASSERT(result, result.message());
+  // Immediately fail the connection.
+  failover_xds_stream_->finishGrpcStream(Grpc::Status::Internal);
+
+  // Ensure that Envoy still attempts to connect to the primary,
```

Comment:
aren't we are ensuring that primary connection is not attempted? (since `envoy.reloadable_features.xds_failover_to_primary_enabled` is set to `false` and we have previously successfully connected to failover)

Reply:
Oops.. the comment was wrong.
Thanks, fixed.